### ISSUE TITLE
Choose the "default editor" per repository.

### ISSUE THREAD

**Issue by struffel** [desktop/desktop#12195]
### Describe the feature or problem you’d like to solve

I have several repositories in GitHub Desktop. Many of them contain code which I like to edit in Visual Studio Code. Others contain Markdown files which I want to edit in Typora. However, at the moment I can only set one external editor to be used for all repositories.

### Proposed solution

Add the ability to select another editor for a specific repository, similar to the way the Git config is handled.
### Additional context

For reference, here is a screenshot of the Git config page in the settings for one repository.
![image](https://user-images.githubusercontent.com/31403260/117870145-04cb3280-b29c-11eb-8a34-9ec941531b64.png)

This would be a good place to add an additional section called "editors" where I can either choose to use the global default editor or select another one just for this repository.

**Comment by steveward** [comment]
Thanks for the issue @StruffelProductions. This was previously requested in https://github.com/desktop/desktop/issues/7621, but since then we haven't heard any other feedback about this. I'll leave this issue open for future discussion and consideration.

**Comment by Rexogamer** [comment]
I can easily see use cases for this. Say you had the JetBrains series of editors, which are usually designed individually for a specific language - it'd be helpful to have them set individually. It's certainly not ideal to go into the app's settings and change the editor every time you switch repos.

**Comment by itamarc** [comment]
Even with the same language it's possible to need a different editor for a specific repository. In some of my Java projects I use VSCode, but for a particular project where I have a Swing interface, I use NetBeans. If I want to develop an Android app, I'll probably use Android Studio.

I've noticed also that NetBeans does not show up in the "External editor" selector in the Options. It's important also to have some way to add a new editor to the list.

**Comment by goyalyashpal** [comment]
this is duplicate of https://github.com/desktop/desktop/issues/10321 i guess, and if so, then this issue should be closed

**Comment by goyalyashpal** [comment]
> - I've noticed also that NetBeans does not show up in the "External editor"
> - It's important also to have some way to add a new editor to the list.

> _- https://github.com/desktop/desktop/issues/12195#issuecomment-901949687_

hi, as per [configuring-a-default-editor#supported-editors](https://docs.github.com/en/desktop/installing-and-configuring-github-desktop/configuring-and-customizing-github-desktop/configuring-a-default-editor#supported-editors), yes, NetBeans is not listed in supported there as of today. I'd say that please open this as a new issue if it doesnt exist already.

<img src="https://user-images.githubusercontent.com/19423063/140878451-1ec99b20-e935-4a0b-b29d-2a22e067d6d3.png" height="320px">

The above link found at: https://github.com/desktop/desktop/issues/10502#issuecomment-687035726

**Comment by steveward** [comment]
Requested in https://github.com/desktop/desktop/issues/13593.

**Comment by steveward** [comment]
Requested in https://github.com/desktop/desktop/issues/14749.

**Comment by steveward** [comment]
Requested in https://github.com/desktop/desktop/issues/14809.

**Comment by moritztim** [comment]
I could also imagine it getting selected automatically by checking the default editor for the most used language or checking for editor traces like `.vscode/` or `.sln`

**Comment by moritztim** [comment]
> I could also imagine it getting selected automatically by checking the default editor for the most used language or checking for editor traces like `.vscode/` or `.sln`

of course it would still be customisable

**Comment by steveward** [comment]
Requested in https://github.com/desktop/desktop/issues/15662.

**Comment by wiverson** [comment]
This is going to be bigger deal as more projects get split between IDEs and users. For example, one dev might be using WebStorm, another VS Code, and another using Fleet all for the same repo.

I would suggest adding all external editors in the menu bar first (as this is AFAIK only a UI change, no settings need to change).

#15662 was closed as duplicate, but I would suggest that those are two different issues/fixes. One (adding default editor per repository) presumably requires sorting out persistence, multi-user scenarios, etc. 

#15662 is (hopefully) a much easier fix.

**Comment by steveward** [comment]
Requested in https://github.com/desktop/desktop/issues/16236.

**Comment by JenuelDev** [comment]
This would be really important,, I can see the external editors in `File > Options > Integrations`. But its for default. It would be awesome if there is an option to select an editor in `Repository > Repository Settings ` that way we can then select an editor for specific repo.

I also really need this one, because currently I have to click folder, and open PhpStorm on that folder. It would be awesome if I can set editor for each of my repo. thanks thanks

**Comment by sweezyio** [comment]
I think for the sake of consistency there should also be the option to change the default shell on a per-repository basis as well.

From some digging it looks it is relatively straightforward to add a new tab to the Repository Settings dialog in [`repository-settings.tsx`](https://github.com/desktop/desktop/blob/development/app/src/ui/repository-settings/repository-settings.tsx), then from there creating a file to describe the settings in [`./repository-settings`](https://github.com/desktop/desktop/blob/development/app/src/ui/repository-settings/). 
<details>
<summary>Notes</summary>
 
- I think we could use `getAvailableEditors()` and `getAvailableShells()` from [`lookup.ts`](https://github.com/desktop/desktop/blob/development/app/src/lib/editors/lookup.ts) and [`shared.ts`](https://github.com/desktop/desktop/blob/development/app/src/lib/shells/shared.ts), respectively, to get the editors and shells to feed into the dialog.
- We can use the same format as the Git config page in the Repository Settings dialog whereby Desktop has the default as the global config and users have the option to use a local config instead.
</details>



The part I am not so sure about is actually saving the settings. From my understanding, the settings are stored within [`app-store.ts`](https://github.com/desktop/desktop/blob/development/app/src/ui/repository-settings/repository-settings.tsx), but currently the editor and shell integrations seem to be global. As a result of this I am unsure on how to proceed with this feature.

**Comment by tidy-dev** [comment]
Requested in #16564 


> Save "External Editor" for every Repository replace Global config
Some project using Android Studio and some using VS Code or Webstorm, we lost more time to change this by click Option and select Another editor


> Save config in .ghdesktop file for every repository, or another solution

**Comment by steveward** [comment]
Requested in https://github.com/desktop/desktop/issues/16690.

**Comment by JenuelDev** [comment]
any update for this one,, I am about to create this suggestion when I saw this.. hehe

**Comment by CosmicElysium** [comment]
This would be a very useful feature. GitHub Desktop is difficult to use with Jetbrains IDEs as my different GitHub projects rely on their different IDEs. A smarter "file association" feature would be ideal (open .cpp files with CLion, .cs files with Rider, etc.) but a per repo setting would also be essential.

**Comment by surkaa** [comment]
我很支持此功能, 在遇到文件冲突需要双击文件在编辑器修改时打开的是默认选中的编辑器, 如果没有这个功能, 我可能会遇到以下问题: 使用Clion打开kotlin文件或者其他更离谱的

I strongly support this feature. When encountering file conflicts and needing to double-click a file to open it in the editor for modification, it is important that the default selected editor is opened. Without this feature, I might encounter the following issues: when using CLion to open Kotlin files or other more unconventional files.

**Comment by brandon-nichols-phd** [comment]
Cmon man just do it

**Comment by mon-jai** [comment]
+1 for this.

**Comment by mon-jai** [comment]
+1 for this.

You won't do Android development in VSCode or web development in Android Studio.

Each repo has its own set of development requirements that need to be addressed with different IDEs.

**Comment by JenuelDev** [comment]
any update on this? or progress?

**Comment by hulsbo** [comment]
+1 on this issue. Also greater support for editors, in my case support for OxygenXML would be neat.

**Comment by tidy-dev** [comment]
requested in https://github.com/desktop/desktop/issues/18027

**Comment by sergiou87** [comment]
Requested in #18128

**Comment by steveward** [comment]
Requested in https://github.com/desktop/desktop/issues/18161.

> I would love to be able to assign an external editor specifically to any specific repository that would override the main default external editor but only for the chosen repository.
> 
> I work a lot in Visual Studio Code, which is why it is my chosen external editor; however, sometimes I work on a Java project such as a Minecraft mod, for which I use IntelliJ IDEA or other types of projects requiring different IDEs. In cases like this, I would love to open my project easily in IDEA (as an example) with the GitHub Desktop shortcut while also being able to open other repositories in something else.
> 
> TL;DR: It would be amazing for specific repositories to be able to set up a different external editor to open the repository.
>

**Comment by tidy-dev** [comment]
Requested in #18163 


> The feature request
Title: Automatically detects the most-appropriate IDE for current repository to call by CMD+SHIFT+A.
>
> Description: Different repository holds different projects. Currently, GitHub Desktop app only allows users to specificy only one IDE to open current repository by CMD+SHIFT+A.
>
>Proposed solution
Suppose that you have different repositories managing an app's Xcode project and homepage, and he wants to use Xcode to open the Xcode project (and VSCode for the homepage repo). In this case, you want this feature request to be considered.
>
>Additional context
N/A

**Comment by steveward** [comment]
Requested in https://github.com/desktop/desktop/issues/18293.

**Comment by RobertLupas** [comment]
Will anybody work on this?
I'd need this feature and would be up to trying to implement it myself. But is it worth it for others, and would it get accepted and pulled into main?

**Comment by steveward** [comment]
From https://github.com/desktop/desktop/issues/18894:

>Desktop has a nice feature that allows you to set your preferred external editor and shell. However, this preference is at the root level and applies to all repositories. It would be really nice if this was a repository-level setting so I could, for example, have VS Code as my preferred editor for some repos and PyCharm as my preferred editor for my Python-based repositories.

**Comment by RobertLupas** [comment]
#18737

**Comment by parth-shpmts** [comment]
![luluwpp](https://github.com/user-attachments/assets/f73f89c1-4b92-4139-83ef-4c9c71538d17)

**Comment by sijanstu** [comment]
Is there any update? already 5 years with no such feature.

**Comment by JenuelDev** [comment]
did someone created a PR for this one?

**Comment by PwiJason** [comment]
2025 and still just a wish/dream

**Comment by swapnilkha** [comment]
Any update on this?

**Comment by JenuelDev** [comment]
did someone made a pr for this?

**Comment by cecell** [comment]
+1 y'all. This is a pretty big missing feature actually. One person may very well be working on two different projects that are for two different people that have different practices / requirements. I ask you kindly to not brush off the issue with the logic that such a person would be using a more robust and professional option. You're better than that...I think

**Comment by ItsChrisHickman** [comment]
What cecell just stated is VERY true.  Especially now with different security policies concerning IDEs that use AI -- VC + CoPilot for some, and Cursor for personal projects --- it's annoying to have to keep swapping out the default IDE to reopen things.

This seems like a rather simple feature -- even simply an 'Open With...'

**Comment by AaronBakerBio** [comment]
This is the most annoying part of gh desktop. IT is super annoying when I am doing a python project and a js project at the same time to not have the ability to set it to use one editor or another. I spend a ton of wasted time between projects trying to remember the editor switch.

**Comment by iamprincejkc** [comment]
no update ?

**Comment by cainsrychl** [comment]
> Requested in [#13593](https://github.com/desktop/desktop/issues/13593).

Not sure if you guys #vscode team punks are stupid or what. Its been 5 fucking years, how about your add this feature, or at the very LEAST post an update. Stop merging requests fuckers.

**Comment by cainsrychl** [comment]
> I can easily see use cases for this. Say you had the JetBrains series of editors, which are usually designed individually for a specific language - it'd be helpful to have them set individually. It's certainly not ideal to go into the app's settings and change the editor every time you switch repos.

If you really see the use case dotard, go fucking add it then.

**Comment by cainsrychl** [comment]
> this is duplicate of [#10321](https://github.com/desktop/desktop/issues/10321) i guess, and if so, then this issue should be closed

If you really see the use case, go add it then. @steveward

**Comment by ADecametre** [comment]
### Temporary solution
Until they develop a solution in 2035, I made this simple script:
```cmd
cd %1
IF EXIST .vs (
	for %%f in (*.sln) do (
		start devenv %%f
	)
) ELSE IF EXIST .idea (
	idea .
) ELSE (
	code .
)
```
<sup>It changes directory to the target path (`%1`), then checks if the current directory contains `.vs` or `.idea`, and opens it with Visual Studio, IntelliJ IDEA or Visual Studio Code.</sup>
1. Create a file ending in `.cmd`
2. Right-click on the file and click on _Copy as path_
3. In GitHub Desktop, go to `File > Options… > Integrations`
  a. External editor: `Configure custom editor…`
  b. Path: `C:\Windows\System32\cmd.exe`
  c. Arguments: `/c "the/path/you/copied.cmd" %TARGET_PATH%`
4. In the `.cmd` file, paste the code above, and add your own if-else statements and code editors

Now GitHub Desktop will open a different editor for each repository.

**Comment by cecell** [comment]
As an absolute nobody, I nominate you as a candidate for serious consideration by Github HR.
https://www.github.careers/careers-home

> ### Temporary solution
> Until they develop a solution in 2035, I made this simple script:
> 
> cd %1
> IF EXIST .vs (
> 	for %%f in (*.sln) do (
> 		start devenv %%f
> 	)
> ) ELSE IF EXIST .idea (
> 	idea .
> ) ELSE (
> 	code .
> )
> It changes directory to the target path (`%1`), then checks if the current directory contains `.vs` or `.idea`, and opens it with Visual Studio, IntelliJ IDEA or Visual Studio Code.
> 
> 1. Create a file ending in `.cmd`
> 2. Right-click on the file and click on _Copy as path_
> 3. In GitHub Desktop, go to `File > Options… > Integrations`
>    a. External editor: `Configure custom editor…`
>    b. Path: `C:\Windows\System32\cmd.exe`
>    c. Arguments: `/c "the/path/you/copied.cmd" %TARGET_PATH%`
> 4. In the `.cmd` file, paste the code above, and add your own if-else statements and code editors
> 
> Now GitHub Desktop will open a different editor for each repository.

**Comment by sweezyio** [comment]
Requested in #21198.

**Comment by pol-rivero** [comment]
For anyone looking for this feature, this was implemented a while ago in my fork: [pol-rivero/github-desktop-plus](https://github.com/pol-rivero/github-desktop-plus).

**Comment by jackfreem** [comment]
Following up on the recently merged "Open With..." feature (#21436), I'd like to propose adding the ability to save editor preferences per-repository.

  How it would work:

  - The existing "Open With..." dialog would get a new checkbox: "Remember this choice for this repository"
  - When checked, the selected editor becomes the default for that specific repository
  - The "Open in {Editor}" menu item would dynamically show the per-repo editor if one is set, otherwise the global default
  - A new "Editor" tab in Repository Settings would allow users to view/clear the per-repo preference

  Storage:

  Per @niik's feedback on #21297, this would extend IDatabaseRepository directly rather than using WorkflowPreferences, since editor preferences are conceptually separate from contribution workflow settings.


  Looking for feedback on this approach before moving forward with implementation @desktop/engineering

**Comment by cecell** [comment]
This app as it is, is something Github likely wishes they could kill off or at least abandon. In its current state, this app generates minimal if any return. They're likely working on something behind the scenes that integrates Copilot and will be offered as a subscription service. For now, they would prefer that we use an IDE that supports Copilot integration and thus a potential source of licensing royalties/billable income. Go make the necessary effort to learn to use a real IDE. In most you'll be able to pick Copilot, or not. Don't worry though, the time you spend learning another app's features won't be a waste. In today's software development climate, if they plan on releasing their own IDE, they'll spend years embroiled in patent fights and ultimately settle out of court with various IP owners in order to provide the same basic features present in every other IDE.

**Comment by adesai1000** [comment]
I have this working in my fork, GlitchHub Desktop. Repository Settings has a new Editor tab with a dropdown: "Use default" or any editor installed on the machine. Once set, every "Open in editor" button, the repository context menu and the app menu use that editor for that repository, while everything else keeps the global choice from Settings.

Not affiliated with GitHub. macOS on Apple silicon only at the moment, and the build is unsigned, so on first launch right-click the app and choose Open. It keeps a separate profile folder from GitHub Desktop.

Download: https://github.com/adesai1000/GlitchHub-Desktop/releases/latest

A star on the repo helps others find it if it solves this for you.
